### PR TITLE
WIP: 0.4 compatibility fixes

### PR DESCRIPTION
**Pull request by stevengj**
This eliminates the deprecation warnings etc. in 0.4 via the Compat package or simple transformations.

The tests pass in 0.3.7 on my machine.  Unfortunately, I'm still getting a segfault in 0.4; I'm not sure what the problem is.
```
signal (11): Segmentation fault: 11
after_read at /Users/stevenj/.julia/HDF5/src/JLD.jl:369
read_scalar at /Users/stevenj/.julia/HDF5/src/JLD.jl:359
jl_apply at /Users/stevenj/Code/julia/src/gf.c:1693
read at /Users/stevenj/.julia/HDF5/src/JLD.jl:331
jl_apply at /Users/stevenj/Code/julia/src/./julia.h:1261
...
```

### CODE REVIEW

**Comment by timholy** [comment]
It seems you're not alone in that segfault. I can reproduce it with this snippet from our test scripts:
```julia
using HDF5, JLD

fn = joinpath(tempdir(),"test.jld")
syms = [:a, :b]
d = Dict([(syms[1],"aardvark"), (syms[2], "banana")])
jldopen(fn, "w") do file
    write(file, "d", d)
end

jldopen(fn) do file
    read(file, "d")
end
```
I bet this is since the `cfunction`/`ccall` ABI changes.

**Comment by simonster** [comment]
The segfault should be fixed by 5b60a15de9957ae332a6cc2561a1b8e08c11df0a.

**Comment by timholy** [comment]
Thanks, @simonster! In 4b63f1a261112fcbefe043fbc9514f1969f78d07 I fixed the next error; the tests now complete, but with several deprecation warnings. @stevengj, perhaps you want to rebase on current master?

**Comment by stevengj** [comment]
Okay, rebased.

**Comment by stevengj** [comment]
Note that there was a warning `Warning: using DataFrames.save in module Main conflicts with an existing identifier.`  I worked around this by using `import DataFrames` instead of `using`, but it would be good to have a better solution for preventing the DataFrames and JLD from conflicting.

**Comment by stevengj** [comment]
Hmm, the test passed once, then I rebuilt HDF5 and I'm getting `ERROR: LoadError: LoadError: AssertionError: HDF5.isvalid(dt)` at `test/gc.jl:25`

File: src/JLD.jl [hunk 1]
```diff
@@ -9,12 +9,12 @@ import HDF5: close, dump, exists, file, getindex, setindex!, g_create, g_open, o
              HDF5ReferenceObj, HDF5BitsKind, ismmappable, readmmap
 import Base: length, endof, show, done, next, ndims, start, delete!, size, sizeof
 
-# .jld files written before v"0.4.0-dev+1419" might have Uint32 instead of UInt32 as the typename string.
+# .jld files written before v"0.4.0-dev+1419" might have UInt32 instead of UInt32 as the typename string.
```
Comment (timholy) [inline src/JLD.jl]: Over-zealous on search-replace here.

**Comment by timholy** [comment]
Looks very nice. I'll merge this (and fix the comment myself in some other commit). It will make it easier to debug any remaining problems.

**Comment by timholy** [comment]
And of course: thanks!

**Pull request merged**